### PR TITLE
git_extensions: refactor and delete redundant functions

### PR DESCRIPTION
**Pull request by jonchang**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

We have some duplicated functionality in utils/git.rb that should be instead in extend/git_extensions.rb which provides a much more expressive API.

This will also help ensure that our code dealing with upstream remotes is agnostic to the choice of default branch.

In general, I think features in git_extensions.rb should be things that deal directly with git repositories. So anything in utils/git.rb that takes a repository argument should instead be in git_extensions.rb. There's still a few more in there but I wanted to make these trickier migrations first since the origin_branch change is pretty invasive, but all other functions remaining only have a single consumer in the codebase.

### CODE REVIEW

**Comment by reitermarkus** [comment]
Are the old `Utils::Git` methods used by any external commands?

File: Library/Homebrew/extend/git_repository.rb [hunk 1]
```diff
@@ -1,53 +1,95 @@
-# typed: false
+# typed: strict
 # frozen_string_literal: true
 
 require "utils/git"
 require "utils/popen"
 
+# Extensions to {Pathname} for querying Git repository information.
+# @see Utils::Git
```
Comment (reitermarkus) [inline Library/Homebrew/extend/git_repository.rb:8]: `@api private`

**Comment by jonchang** [comment]
> Are the old `Utils::Git` methods used by any external commands?

That's a good point. I'll add the old methods back in and do an `odeprecated`, and also check Homebrew org taps for usage of these.

**Comment by jonchang** [comment]
I checked all the Homebrew org taps and didn't find any other usage of these functions. I'll still use the deprecation notice just in case a third-party tap relies on them though.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 2]
```diff
@@ -84,44 +84,38 @@ def bump_formula_pr_args
...
-          if quiet_system "git", "cat-file", "-e", "#{origin_branch}:#{formula_path}"
-            ohai "#{formula.full_name} exists in #{origin_branch}"
-            safe_system "git", "checkout", origin_branch
-            return tap_full_name, origin_branch, previous_branch
-          end
-        end
-      end
+    default_origin_branch = formula.tap.path.git_origin_branch if formula.tap
+
+    return formula.tap&.full_name, "origin", default_origin_branch, "-" unless OS.linux? && formula.tap.core_tap?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:89]: Can this be refactored to not use `unless .. &&` which is hard to parse.

**Review by MikeMcQuaid — APPROVED** [review]
👍🏻 once 🟢
